driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,280 @@
+    private static final String MONGODB_IAM_MECHANISM_NAME = "MONGODB-IAM";
+    private static final int RANDOM_LENGTH = 32;
+
+    public IamAuthenticator(final MongoCredentialWithCache credential) {
+        super(credential);
+
+        if (getMongoCredential().getAuthenticationMechanism() != MONGODB_IAM) {
+            throw new MongoException("Incorrect mechanism: " + getMongoCredential().getMechanism());
+        }
+    }
+
+    @Override
+    public String getMechanismName() {
+        return MONGODB_IAM_MECHANISM_NAME;
+    }
+
+    @Override
+    protected SaslClient createSaslClient(final ServerAddress serverAddress) {
+        return new IamSaslClient(getMongoCredential());
+    }
+
+    private static class IamSaslClient implements SaslClient {
+        private final MongoCredential credential;
+        private int step = -1;
+        private byte[] clientNonce = new byte[RANDOM_LENGTH];
+        private String httpResponse;
+
+        IamSaslClient(final MongoCredential credential) {
+            this.credential = credential;
+        }
+
+        @Override
+        public String getMechanismName() {
+            AuthenticationMechanism authMechanism = credential.getAuthenticationMechanism();
+            if (authMechanism == null) {
+                throw new IllegalArgumentException("Authentication mechanism cannot be null");
+            }
+            return authMechanism.getMechanismName();
+        }
+
+        @Override
+        public boolean hasInitialResponse() {
+            return true;
+        }
+
+        @Override
+        public byte[] evaluateChallenge(final byte[] challenge) throws SaslException {
+            step++;
+            if (step == 0) {
+                return computeClientFirstMessage();
+            }
+            if (step == 1) {
+                return computeClientFinalMessage(challenge);
+            } else {
+                throw new SaslException(format("Too many steps involved in the %s negotiation.", getMechanismName()));
+            }
+        }
+
+        @Override
+        public boolean isComplete() {
+            return step == 1;
+        }
+
+        @Override
+        public byte[] unwrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public byte[] wrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public Object getNegotiatedProperty(final String s) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public void dispose() throws SaslException {
+            // nothing to do
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            new SecureRandom(this.clientNonce).nextBytes(this.clientNonce);
```

Comment:
Don't pass a seed to `SecureRandom`. You will may make the random numbers predictable

Reply:
Removed.